File: examples/python/Camera/camera_hdr.py
#!/usr/bin/env python3

import cv2
import depthai as dai

# HDR Camera example
# Use 'h' and 'j' keys to increase/decrease HDR exposure ratio
# Use 'w' and 'e' keys to increase/decrease local tone weight
# Use 'b' to toggle HDR exposure base between 'long' and 'middle'
# Use 'q' to quit

# This example only works on IMX586/582 sensors.

with dai.Pipeline() as pipeline:
    cam = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_A)
    cameraFeatures = pipeline.getDefaultDevice().getConnectedCameraFeatures()
    success = False
    for feature in cameraFeatures:
        if feature.socket == dai.CameraBoardSocket.CAM_A:
            if feature.sensorName in ["IMX586", "IMX582", "LCM48"]:
                success = True
                break

    if not success:
        raise RuntimeError("No suitable camera found, HDR is only supported on IMX586/582 sensors!")

    cam.initialControl.setHdr(True)
    cameraControlQueue = cam.inputControl.createInputQueue()
    videoQueue = cam.requestOutput((1280,800)).createOutputQueue()

    minRatio = 1
    maxRatio = 8
    ratio = 4

    minWeight = 0.0
    maxWeight = 1.0
    weight = 0.75

    base = "long"

    pipeline.start()

    while pipeline.isRunning():
        def clamp(n, smallest, largest):
            return max(smallest, min(n, largest))
        videoIn = videoQueue.get()
        assert isinstance(videoIn, dai.ImgFrame)
        cv2.imshow("video", videoIn.getCvFrame())
        key = cv2.waitKey(1)

        ctrl = dai.CameraControl()
        if key == ord("h"):
            ratio = clamp(ratio * 2, minRatio, maxRatio)
            ctrl.setMisc("hdr-exposure-ratio", ratio)
            print(f"Setting HDR exposure ratio to {ratio}")
            cameraControlQueue.send(ctrl)
        if key == ord("j"):
            ratio = clamp(ratio // 2, minRatio, maxRatio)
            ctrl.setMisc("hdr-exposure-ratio", ratio)
            print(f"Setting HDR exposure ratio to {ratio}{' (off)' if ratio == 1 else ''}")
            cameraControlQueue.send(ctrl)
        if key == ord("w"):
            weight = clamp(weight + 1/32, minWeight, maxWeight)
            ctrl.setMisc("hdr-local-tone-weight", weight)
            print(f"Setting HDR local tone weight to {weight}")
            cameraControlQueue.send(ctrl)
        if key == ord("e"):
            weight = clamp(weight - 1/32, minWeight, maxWeight)
            ctrl.setMisc("hdr-local-tone-weight", weight)
            print(f"Setting HDR local tone weight to {weight}")
            cameraControlQueue.send(ctrl)
        if key == ord("b"):
            base = ("middle" if base == "long" else "long")
            ctrl.setMisc("hdr-exposure-base", base)
            print(f"Setting HDR exposure base to: {base}")
            cameraControlQueue.send(ctrl)
        if key == ord("q"):
            break

File: examples/cpp/Camera/camera_hdr.cpp
#include <iostream>
#include <memory>
#include <opencv2/opencv.hpp>

#include "depthai/common/CameraBoardSocket.hpp"
#include "depthai/depthai.hpp"

/*
    HDR Camera example
    Use 'h' and 'j' keys to increase/decrease HDR exposure ratio
    Use 'w' and 'e' keys to increase/decrease local tone weight

    This example only works on IMX586/582 sensors.
*/

float clamp(float n, float smallest, float largest) {
    return std::max(smallest, std::min(n, largest));
}

int main() {
    // Create pipeline
    dai::Pipeline pipeline;

    // Define source and output
    auto cam = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_A);
    auto cameraFeatures = pipeline.getDefaultDevice()->getConnectedCameraFeatures();
    bool success = false;
    for(const auto& feature : cameraFeatures) {
        if(feature.socket == dai::CameraBoardSocket::CAM_A) {
            if(feature.sensorName == "IMX586" || feature.sensorName == "IMX582" || feature.sensorName == "LCM48") {
                success = true;
                break;
            }
        }
    }

    if(!success) {
        throw std::runtime_error("No suitable camera found, HDR is only supported on IMX586/582 sensors!");
    }
    cam->initialControl.setHdr(true);
    auto cameraControlQueue = cam->inputControl.createInputQueue();
    auto videoQueue = cam->requestOutput({1280, 800})->createOutputQueue();

    // Initialize variables
    float minRatio = 1.0f;
    float maxRatio = 8.0f;
    float ratio = 4.0f;

    float minWeight = 0.0f;
    float maxWeight = 1.0f;
    float weight = 0.75f;

    std::string base = "long";

    // Start pipeline
    pipeline.start();
    while(pipeline.isRunning()) {
        auto videoIn = videoQueue->get<dai::ImgFrame>();
        if(videoIn == nullptr) continue;

        cv::imshow("video", videoIn->getCvFrame());
        int key = cv::waitKey(1);
        auto ctrl = std::make_shared<dai::CameraControl>();

        // Key handling
        if(key == 'h') {
            ratio = clamp(ratio * 2.0f, minRatio, maxRatio);
            ctrl->setMisc("hdr-exposure-ratio", ratio);
            std::cout << "Setting HDR exposure ratio to " << ratio << std::endl;
            cameraControlQueue->send(ctrl);
        } else if(key == 'j') {
            ratio = clamp(ratio / 2.0f, minRatio, maxRatio);
            ctrl->setMisc("hdr-exposure-ratio", ratio);
            std::cout << "Setting HDR exposure ratio to " << ratio << std::endl;
            cameraControlQueue->send(ctrl);
        } else if(key == 'w') {
            weight = clamp(weight + 1/32.0f, minWeight, maxWeight);
            ctrl->setMisc("hdr-local-tone-weight", weight);
            std::cout << "Setting HDR local tone weight to " << weight << std::endl;
            cameraControlQueue->send(ctrl);
        } else if(key == 'e') {
            weight = clamp(weight - 1/32.0f, minWeight, maxWeight);
            ctrl->setMisc("hdr-local-tone-weight", weight);
            std::cout << "Setting HDR local tone weight to " << weight << std::endl;
            cameraControlQueue->send(ctrl);
        } else if(key == 'b') {
            base = (base == "long") ? "middle" : "long";
            ctrl->setMisc("hdr-exposure-base", base);
            std::cout << "Setting HDR exposure base to '" << base << "'" << std::endl;
            cameraControlQueue->send(ctrl);
        } else if(key == 'q') {
            break;
        }
    }

    return 0;
}